src/bodywork/workflow_execution.py

```diff
@@ -109,6 +97,7 @@ def run_workflow(
         env_vars = k8s.create_k8s_environment_variables(
             [(GIT_COMMIT_HASH_K8S_ENV_VAR, get_git_commit_hash(cloned_repo_dir))]
         )
+        workflow_has_services = False
```

Comment:
We could answer this directly via `config`?

I have an aversion to mutating variables like this, because it makes the code harder to follow.

Reply:
I know what you mean but I think the config solution will mean creating another nested for loop just for this as you need to check both the DAG and the stage list.

Reply:
```python
def config_deploys_services(config: BodyworkConfig) -> bool:
    service_stages = [
        stage_name
        for stage_name, stage in config.stages.items()
        if isinstance(stage, ServiceStageConfig) and stage_name in config.project.DAG
    ]
    deploys_service_stages = True if service_stages else False
    return deploys_service_stages
```

Reply:
My point exactly!  A whole method and repetition of code for something that is covered by one bool. I can't agree that this is a more optimal solution in this instance.

Reply:
Also it's just one bool in the whole method.

Reply:
Yeah, but you re-assign a value to the dummy variable, every time you hit that part of the code (i.e. for every step), which is far from elegant. At least `config_deploys_services` is just one-pass of the config file - it's just one list comprehension.

Reply:
Assigning to the bool again isn't optimal but it is far more efficient than iterating through the entire list again as assigning to a bool is a very cheap operation.  Also it takes fewer lines of code. 

What annoys me is that Python misses C# LINQ functions, specifically to use an Any() function with my parameters. Thinking about how that works I think the following is a compromise whilst still being fairly efficient:

`def _deploys_services(config: BodyworkConfig) -> bool:
    for stage_name, stage in config.stages.items():
        if isinstance(stage, ServiceStageConfig) and stage_name in config.project.DAG:
            return True
    return False`

Reply:
Assigning to the bool again isn't optimal but it is far more efficient than iterating through the entire list again as assigning to a bool is a very cheap operation. Also it takes fewer lines of code.

What annoys me is that Python misses C# LINQ functions, specifically to use an Any() function with my parameters. Thinking about how that works I think the following is a compromise whilst still being fairly efficient:

```python
def _deploys_services(config: BodyworkConfig) -> bool:
    for stage_name, stage in config.stages.items():
        if isinstance(stage, ServiceStageConfig) and stage_name in config.project.DAG:
            return True
    return False
    ```

Reply:
There is an `any`:

```python
def config_deploys_services(config: BodyworkConfig) -> bool:
    return any(True for stage_name, stage in config.stages.items()
        if isinstance(stage, ServiceStageConfig) and stage_name in config.project.DAG
    )
```

?

Reply:
We have a winner! I didn't think we could use it like that from the docs but that's how I wanted to use it although the syntax is different.

Reply:
💥